### ISSUE TITLE
When only active transform stream exceeds highWaterMark, process exits

### ISSUE THREAD

**Issue by joepie91** [nodejs/node-v0.x-archive#14477]
Node.js version: `v0.12.0`

Test case:

``` javascript
var fs = require("fs");
var stream = require("stream");

var transform = new stream.Transform({
  highWaterMark: 200 * 1024
});

transform._transform = function(chunk, encoding, callback) {
  console.log("Got chunk, length", chunk.length);
  this.push(chunk);
  return callback();
};

console.log("Starting stream...");

fs.createReadStream("./lower.txt")
  .pipe(transform)
  .on("end", function() {
      console.log("Done!");
  });
```

Expected result: Buffer fills up, process/stream stalls until buffer is empty enough again.

Actual result: Process exits without ever completing the stream, and without throwing any kind of error.

```
sven@linux-etoq:~/projects/node-bhttp> coffee test-exit.coffee 
Starting stream...
Got chunk, length 65536
Got chunk, length 65536
Got chunk, length 65536
Got chunk, length 65536
sven@linux-etoq:~/projects/node-bhttp> 
```

I presume that this is because the stream is paused and the event loop runs out of events, but it's very unintuitive, and it took me a good while to debug when I ran into it.

**Comment by misterdjules** [comment]
Right, this issue is specific to using a `fs.ReadStream`. These streams are implemented in terms of libuv requests. For each chunk of data that is read, there is a corresponding libuv request. Each of these request completes whenever the chunk of data is read, and a new request is created/started when the previous one completes.

Thus as you said, if there's any pause in the streaming process, it is possible right after request `n-1` completes and right before request `n` kicks in, that no libuv request is active. If that's the only I/O in progress, node will exit.

Using the same code but piping from a socket to the same transform stream that uses the same `highWaterMark` setting would work as expected, because network I/O uses libuv handles, which are active from the beginning until the end of the whole I/O process.

Here's some sample code that illustrates this:
````
var net = require('net');
var stream = require("stream");
var util = require('util');

function MyTransformStream(options) {
  stream.Transform.call(this, options);
}
util.inherits(MyTransformStream, stream.Transform);

MyTransformStream.prototype._transform = function _transform(chunk, encoding, callback) {
  console.log("Got chunk, length", chunk.length);
  this.push(chunk);
  return callback();
};

var transformStream = new MyTransformStream({
    highWaterMark: 200 * 102
});

transformStream.on('end', function onEnd() {
  console.log('Got end on transform stream!');
});

// Start reading on the transform stream so that
// the pipeline is kept flowing
transformStream.on('data', function() {});

var server = net.createServer(function onConnection(socket) {
  // Make sure that libuv's loop is not kept alive by the server's handle
  server.unref();
  socket.pipe(transformStream);
});

server.listen(4242, function() {
  console.log('server listening...');
})
````
Running that and using `nc` to connect and send some data:
```
dd if=/dev/random bs=100k count=100 | nc localhost 4242
```
we can see the `'Got end on transform stream!'` message.

I would suggest renaming the issue and changing the description to make it more obvious this is an issue with `fs.ReadStream`.

**Comment by joepie91** [comment]
@misterdjules The same thing occurs when I modify your example to hit the HWM, though:

```
dd if=/dev/urandom bs=100k count=100 | nc localhost 4242
```

```
sven@linux-etoq:~/projects/node-bhttp> node test-exit2.js 
server listening...
Got chunk, length 1024
Got chunk, length 65536
sven@linux-etoq:~/projects/node-bhttp> 
```

The change I made was to comment out this line:

```javascript
transformStream.on('data', function() {});
```

This is what the issue is really about - I would expect it to either stall upon hitting the HWM, or (less ideally) throw an error.

**Comment by dayuoba** [comment]
@misterdjules i've tried @joepie91's test script,i found a appearance that no matter how big the size of file is ,node will exit after the forth report.
Here is the output
```
Starting stream...
Got chunk, length 65536
Got chunk, length 65536
Got chunk, length 65536
Got chunk, length 65536

```
the file size grows from 1mb to 2.6mb.but when i add a listener `transform.on('data')`,it will run correctly.
I do not think it is accidental like you explained above.

**Comment by misterdjules** [comment]
@joepie91 @dayuoba My analysis and description of the problem was missing important details, thank you for pointing them out :+1:

@joepie91 Indeed, the sample code I mentioned in [my previous comment](https://github.com/joyent/node/issues/14477#issuecomment-90307698) should not have set a listener for the `data` event for the transform stream.

With the sample code that uses the `net.Server` and `nc`, what happens if we don't trigger the flowing mode on the transform stream is that [`readStop` is called on the net.Socket readable stream](https://github.com/joyent/node/blob/master/lib/net.js#L542-L545), and that removes the underlying handle from the active handle at the libuv layer.

If there's nothing else that would keep the libuv loop open, then node exits. I think that's what you described in your EDIT2 @dayuoba. In the case of a network server, that use case would work because most of the time the server handle is kept active, and that at least keeps the libuv loop active.

So the sample code that uses a net.Server and a TCP client fails for similar reasons than the `fs.ReadStream` sample, but the technical details are slightly different.

Basically, node streams do not hold the loop open, even if there's still data to process. I don't know if it's by design, and if so what was the rationale behind this decision. I haven't thought about the pros and cons of both approaches, but making streams hold a loop open or have an underlying handle would certainly not be something to consider lightly.

@joyent/node-coreteam Thoughts about whether this is a bug, or if this works by design? Any recommendation for workarounds?

**Comment by dayuoba** [comment]
@misterdjules in my opinion,it is not a bug,but node also need a 'event' for Buffer full(i am not sure if it had).So we can consume the Buffer when it was full of data.and the stream will go on .

**Comment by jasnell** [comment]
I'm not convinced it's a bug but it's also something that could likely be improved. Perhaps we can mark this as a feature request and revisit later?

**Comment by keatz55** [comment]
+1
